### PR TITLE
Refactor batching server to support coingecko and polygon.io

### LINKED ISSUE #20: Add other price feed providers
We want to add support for different price feed providers, namely:
- fiat prices on Polygon.io
- token prices on Coingecko

The API that our chain consumes should stay exactly the same.

More details can be found in #21.

### PR DESCRIPTION
**Pull request by ebma**
## What changed
- Updates relevant dependencies
- Replaces `structopt` with `clap` for argument parsing
- Adds implementations for querying prices from polygon.io, coingecko, and 'custom' endpoints (for now just the AMPE price from the squid)
- Removes queries to the DIA endpoint
- The `supply` field uses the data of the '24-hour trading volume' field. This aligns to the previous implementation which was using the 24 hour trading volume reported by DIA, though the data apparently differs a lot. This doesn't matter too much as we don't use the `supply` field for anything.
- Simplifies the overall structure and unifies types. 

## How it works
The `PriceApi` in `src/api/mod.rs` is the high-level interface that can be used to query quotations for all supported assets. The `PriceApiImpl` will try to query prices from all supported endpoints. While the `get_quotations()` function returns a `Result`, it **will never return an error**, it will just log any errors that are encountered for any endpoint. I decided to implement it this way because it can always happen that a query fails for some reason and we would not like the batching server to stop because of this. All we can do is log (and possibly also send a reporting message somewhere). 

We are now using a hardcoded mapping in `convert_to_coingecko_id()` in `src/api/coingecko.rs`. We should soon change the `CustomMetadata` of our runtimes, replacing the `diaKeys` field with a new type that allows to directly specify the coingecko identifier. This change would require a runtime upgrade (with a migration) and a root transaction to set the new metadata fields though. For the time being, the hardcoded mapping will require us to touch the code every time we want to add support for a new asset. 

The API keys needed for Polygon and Coingecko can be passed as CLI arguments or via env variables. 

## Notes
While the response from polygon.io contains the timestamp of the last bid/ask placed, see [here](https://polygon.io/docs/forex/get_v1_conversion__from___to), I refrained from using that timestamp and instead used the timestamp of when the query was performed. This is because even if the last bid/ask was placed a while ago (according to the timestamp in the response), this does not necessarily mean that the price info is stale but could mean that trading is closed. 

## Testing

I tested this with the Foucoco runtime in Zombienet and the offchain-worker was able to feed the prices appropriately.

Closes #20.

### CODE REVIEW

File: dia-batching-server/src/api/coingecko.rs [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+		let id_to_price_map =
+			self.client.price(&coingecko_ids, false, true, false, true).await.map_err(|e| {
+				CoingeckoError(format!("Couldn't query CoinGecko prices {}", e.to_string()))
+			})?;
+
+		let quotations = id_to_price_map
+			.into_iter()
+			.filter_map(|(id, price)| {
+				let asset = assets.iter().find(|asset| {
```
Comment (b-yap) [inline dia-batching-server/src/api/coingecko.rs:55]: What's the need to check _again_ if the `id` in _`id_to_price_map`_ is a vaild/existing CoinGecko id?
Wasn't the _`coingecko_ids`_ variable used to generate _`id_to_price_map`_ ?

File: dia-batching-server/src/api/coingecko.rs [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+	pub async fn get_prices(
+		&self,
+		assets: Vec<&AssetSpecifier>,
+	) -> Result<Vec<Quotation>, CoingeckoError> {
+		let coingecko_ids = assets
+			.clone()
+			.into_iter()
+			.filter_map(|asset| {
+				let id = Self::convert_to_coingecko_id(asset);
```
Comment (b-yap) [inline dia-batching-server/src/api/coingecko.rs:36]: Or you can shorten this to:
```rust
.filter_map(|asset| 
	Self::convert_to_coingecko_id(asset).or_else(|| { 
		log::warn!("Could not find CoinGecko ID for asset {:?}", asset);
		None 
	})
)
```

File: dia-batching-server/src/api/coingecko.rs [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+		ids: &[Id],
+		include_market_cap: bool,
+		include_24hr_vol: bool,
+		include_24hr_change: bool,
+		include_last_updated_at: bool,
+	) -> Result<HashMap<String, CoingeckoPrice>, CoingeckoError> {
+		let ids = ids.iter().map(AsRef::as_ref).collect::<Vec<_>>();
+		// We always query for USD
+		let vs_currencies = vec!["usd"];
+		let vs_currencies = vs_currencies.iter().map(AsRef::as_ref).collect::<Vec<_>>();
```
Comment (b-yap) [inline dia-batching-server/src/api/coingecko.rs:188]: `vs_currencies` is already a **`Vec<str>`**. You can remove line 188.

File: dia-batching-server/src/price_updater.rs [hunk 4]
```diff
@@ -25,99 +25,55 @@ where
...
-					Err(err) => {
-						error!("Error while retrieving quotation for {:?}: {}", quoted_asset, err)
-					},
-				}
-			}
-		}
-	}
+	api.get_quotations(supported_currencies)
+		.await
+		.unwrap_or_default()
```
Comment (b-yap) [inline dia-batching-server/src/price_updater.rs:71]: While I understand the reason behind never returning an error, we could also have `fn get_quotation()` to simply return `Vec<Quotation>`.

We can still log the errors with the method returning an empty vec.

File: dia-batching-server/src/api/coingecko.rs [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+					Err(CoingeckoError(err_msg))
+				} else {
+					let result = response.json().await;
+					result.map_err(|e| {
+						CoingeckoError(format!("Could not decode CoinGecko response: {}", e))
+					})
+				}
+			},
+			Err(e) => Err(CoingeckoError(e.to_string())),
+		}
```
Comment (bogdanS98) [inline dia-batching-server/src/api/coingecko.rs:167]: I think we can use `map_err` and get rid of the `match`, something similar to: 
```suggestion
		let response = client.get(url).send().await.map_err(|e| CoingeckoError(e.to_string()))?;
		
		if !response.status().is_success() {
	                let result = response.text().await;
	                let err_msg = format!(
		                "CoinGecko API error: {}",
		                result.unwrap_or("Unknown".to_string()).trim()
	                );
	                return Err(CoingeckoError(err_msg))
                } 
                
                response.json().await.map_err(|e| {
		        CoingeckoError(format!("Could not decode CoinGecko response: {}", e))
	        })				
```

File: dia-batching-server/src/api/coingecko.rs [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+	}
+
+	/// Check API server status
+	#[allow(dead_code)]
+	pub async fn ping(&self) -> Result<SimplePing, CoingeckoError> {
```
Comment (bogdanS98) [inline dia-batching-server/src/api/coingecko.rs:172]: It's nice that they have this endpoint for checking the server status, but how often do we plan to call it and what would be the "retry logic" on our side in case their server is down?

File: dia-batching-server/src/api/polygon.rs [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+	}
+	pub async fn get_prices(
+		&self,
+		assets: Vec<&AssetSpecifier>,
+	) -> Result<Vec<Quotation>, PolygonError> {
+		let asset_and_source_currency = assets
+			.into_iter()
+			.filter_map(|asset| {
+				let source_curreny = PolygonPriceApi::extract_source_currency(asset);
+				match source_curreny {
```
Comment (bogdanS98) [inline dia-batching-server/src/api/polygon.rs:33]:
```suggestion
				let source_currency = PolygonPriceApi::extract_source_currency(asset);
				match source_currency {
```

File: dia-batching-server/src/api/polygon.rs [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+					let result = response.json().await;
+					result.map_err(|e| {
+						PolygonError(
+							"Could not decode Polygon response: ".to_owned() + &e.to_string(),
+						)
+					})
+				}
+			},
+			Err(e) => Err(PolygonError(e.to_string())),
+		}
```
Comment (bogdanS98) [inline dia-batching-server/src/api/polygon.rs:163]: Same as here

File: dia-batching-server/src/api/polygon.rs [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+	}
+
+	/// Get the current price of any fiat currency in USD
+	pub async fn price(&self, from_currency: String) -> Result<PolygonPrice, PolygonError> {
+		// Currencies have to be upper-case
+		let from_currency = from_currency.to_uppercase();
+		// We always query for USD.
+		let to_currency = "USD";
+		// We always query for full precision
+		let precision = "5";
```
Comment (bogdanS98) [inline dia-batching-server/src/api/polygon.rs:173]: Seem like the precision can be at most 4 from Polygon API [docs](https://polygon.io/docs/forex/get_v1_last_quote_currencies__from___to) or am I looking at the wrong doc page? 
Also I cannot query the `conversion` endpoint from their doc page to see what happens when using `5` as the value for `precision` since they require at least a paid Starter subscription to do that.

**Review by bogdanS98 — COMMENTED** [review]
Great changes @ebma 👍🏼, I only left a few minor remarks

File: dia-batching-server/src/api/coingecko.rs (hunk 1 above)
Comment (ebma) [reply dia-batching-server/src/api/coingecko.rs:55]: The 'problem' is that the 'id_to_price_map' does not have information about the `AssetSpecifier`. The `AssetSpecifier` holds information about the `symbol` and `name` that we want to use in the `Quotation` that we return below. The `id_to_price_map` only has Coingecko-related info, ie. the `id` used is the coingecko identifier and the `price` in the map holds the coingecko data. We only have a one-way mapping of `AssetSpecifier` -> `CoingeckoID` so we iterate over all assets in the `id_to_price_map` doing this one-way lookup and trying to find the matching asset again.

File: dia-batching-server/Cargo.toml [hunk 10]
```diff
@@ -3,28 +3,30 @@ name = "dia-batching-server"
 version = "0.1.0"
```
[the remark this answers is not in the document]
Comment (ebma) [reply dia-batching-server/Cargo.toml:3]: I think we should eventually, but not as part of this ticket. There is probably more we should change in that regard (possibly also the other sub-projects) so it's better to do that separately.

File: dia-batching-server/src/api/coingecko.rs (hunk 3 above)
Comment (ebma) [reply dia-batching-server/src/api/coingecko.rs:188]: True, thanks 👍

File: dia-batching-server/src/api/polygon.rs (hunk 9 above)
Comment (ebma) [reply dia-batching-server/src/api/polygon.rs:171]: Yes, that's what they put in their docs but I noticed that I can actually get higher precision by defining a larger value. The highest precision I could get was `5` though. You can even put a precision of `10` or similar but the response will not have more than 5 decimals as far as I can tell. That's why I chose that value here.

File: dia-batching-server/src/api/coingecko.rs (hunk 6 above)
Comment (ebma) [reply dia-batching-server/src/api/coingecko.rs:163]: I added this at first because I thought it would come in handy for testing if the server is down or not but then I realized that it might not be necessary to have that and we can just try to get a response from the usual endpoint and log the error in case something fails. There is nothing to retry or fall back to anyway.

File: dia-batching-server/src/price_updater.rs (hunk 4 above)
Comment (ebma) [reply dia-batching-server/src/price_updater.rs]: I changed the signature of the function to reflect that it doesn't return an Error

**Review by bogdanS98 — APPROVED** [review]
>  I decided to implement it this way because it can always happen that a query fails for some reason and we would not like the batching server to stop because of this. All we can do is log (and possibly also send a reporting message somewhere)

I think that not throwing errors in this case is a good approach. However, if we decide to report errors, we could send them to Prometheus as metrics and then to our Slack channel based on specific alerts that are triggered if a condition is met (e.g. API returns error) over a certain period of time. This method helps distinguish the causes of errors and prevents spamming the Slack channel. For example, we could identify which exact API is failing or if the batching server is down. What do you think?

**Review by gianfra-t — APPROVED** [review]
Lots of changes! I left some comments but if the testing works seems good to me 👍.

File: dia-batching-server/src/api/custom.rs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+pub struct CustomPriceApi;
+
+impl CustomPriceApi {
+	pub async fn get_price(asset: &AssetSpecifier) -> Result<Quotation, CustomError> {
+		if !Self::is_supported(asset) {
+			return Err(CustomError(format!("Unsupported asset: {:?}", asset)));
+		}
+
+		// Go through all the custom APIs and check if the asset is supported by any of them
+		if AmpePriceView::supports(asset) {
```
[the remark this answers is not in the document]
Comment (ebma) [reply dia-batching-server/src/api/custom.rs:19]: It seems redundant but the first check in l. 14 is supposed to check whether the `asset` is generally supported by the `CustomPriceApi` and once we know that it's supported we would go through all the custom APIs to see which one in particular supports that asset. Since we only have the AMPE price as a custom API, it's redundant but this way it's easy to extend with more assets in the future. 

I can also remove the first check and make it iterate over all available custom APIs. Then we can get rid of the definition of `custom_assets` in l. 27. What do you think?

File: dia-batching-server/src/api/polygon.rs [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+					None => {
+						log::warn!("Unsupported polygon asset: {:?}", asset);
+						None
+					},
+				}
+			})
+			.collect::<Vec<_>>();
+
+		let mut prices = Vec::new();
+		for (asset, source_currency) in asset_and_source_currency {
```
[the remark this answers is not in the document]
Comment (ebma) [reply dia-batching-server/src/api/polygon.rs:44]: I think it should be fine. We have an unlimited number of queries available in our plan. Though you do have a point that it might be better to batch the requests with that 'All tickers' endpoint. 

I was curious how much the delay is and one query takes approximately 0.5 seconds on my machine. If we want to support more currencies in the future or run the server on a lower-spec machine, the delay could add up to an undesirable amount. 
```
 INFO  dia_batching_server::api::polygon > Time before fetching prices: 2024-07-02 16:05:05.065909 UTC
 INFO  dia_batching_server::api::polygon > Adding price for TZS-USD at time 2024-07-02 16:05:05.653694 UTC
 INFO  dia_batching_server::api::polygon > Adding price for NGN-USD at time 2024-07-02 16:05:06.276484 UTC
 INFO  dia_batching_server::api::polygon > Adding price for BRL-USD at time 2024-07-02 16:05:06.752161 UTC
 INFO  dia_batching_server::api::polygon > Adding price for AUD-USD at time 2024-07-02 16:05:07.246598 UTC
 INFO  dia_batching_server::api::polygon > Adding price for EUR-USD at time 2024-07-02 16:05:07.842225 UTC
 INFO  dia_batching_server::api::polygon > Time after fetching prices: 2024-07-02 16:05:07.842358 UTC
```

So I think you have a point and I'll try to switch to that other endpoint instead.

File: dia-batching-server/src/api/custom.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ebma) [reply dia-batching-server/src/api/custom.rs]: I refactored it now to make the checks less redundant.

File: dia-batching-server/src/price_updater.rs [hunk 13]
```diff
@@ -25,99 +25,54 @@ where
...
+			tokio::time::sleep(update_interval.saturating_sub(time_elapsed.elapsed())).await;
 		}
 	});
 
 	Ok(())
 }
 
 fn convert_to_coin_info(value: Quotation) -> Result<CoinInfo, Box<dyn Error + Sync + Send>> {
-	let Quotation { name, symbol, blockchain, price, time, volume_yesterday, .. } = value;
+	let Quotation { name, symbol, blockchain, price, time, supply, .. } = value;
```
[the remark this answers is not in the document]
Comment (ebma) [reply dia-batching-server/src/price_updater.rs]: Removed

File: dia-batching-server/src/api/polygon.rs [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+	/// Extract the source currency from the asset pair.
+	/// We assume that the symbol contained in the `AssetSpecifier` is of the form <from>-<to>.
+	fn extract_source_currency(asset: &AssetSpecifier) -> Option<String> {
+		let (blockchain, symbol) = (asset.blockchain.as_str(), asset.symbol.as_str());
+		if blockchain.to_uppercase() != "FIAT" {
+			return None;
+		}
+
+		// We assume to receive a symbol of form <from>-<to> and we want to extract the <from> part
+		let parts: Vec<_> = symbol.split('-').collect();
```
[the remark this answers is not in the document]
Comment (ebma) [reply dia-batching-server/src/api/polygon.rs:93]: Added the check

File: dia-batching-server/src/types.rs [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+/// This struct is used to store information about a coin.
+#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoinInfo {
+	pub symbol: SmolStr,
+	pub name: SmolStr,
+	pub blockchain: SmolStr,
+	pub supply: u128,
```
[the remark this answers is not in the document]
Comment (ebma) [reply dia-batching-server/src/types.rs:35]: Hmm yeah we don't really need to keep it but we already store the supply in our on-chain data of the DIA pallet so removing it now would be a bit tricky. That's why I think it's better to keep it around even if the value is non-sensical for our fiat prices.

File: dia-batching-server/src/api/custom.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (gianfra-t) [reply dia-batching-server/src/api/custom.rs]: Looks great!

File: dia-batching-server/src/api/polygon.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (gianfra-t) [reply dia-batching-server/src/api/polygon.rs:44]: Wow 0.5 seconds seems quite a lot, thanks for testing it. That would mean that supporting 24 currencies would be already too much for a block "resolution". 

But since we are not supporting that many now, maybe you can leave a comment and we can pick this up later as an improvement since it will change the logic a bit, I assume.

[the remark this answers is not in the document]
Comment (ebma) [reply dia-batching-server/src/api/polygon.rs]: I added that batch request now because the results also give a higher precision than just 5 decimals. 

I used the price contained in the `bid` of the last quote because when comparing the price returned by the fiat-conversion endpoint to the tickers, it seemed to match with the value of the `bid` but rounded.

File: dia-batching-server/src/types.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (TorstenStueber) [reply dia-batching-server/src/types.rs:35]: Agreed!

**Comment by ebma** [comment]
I refactored this again to:
- support ENV variables for all configuration parameters
- query fiat prices in a batch instead of single requests
- ignore the capitalization of supplied currencies (all currencies are transformed to and compared with uppercase letters)

I tested it again with the offchain worker on Foucoco zombienet and it looks good to me. @pendulum-chain/devs please have another look.

**Review by bogdanS98 — APPROVED** [review]
LGTM 👍🏼

File: dia-batching-server/src/api/polygon.rs [hunk 16]
```diff
@@ -0,0 +1,393 @@
...
+				time: chrono::Utc::now().timestamp().unsigned_abs(),
+			};
+			prices.push(quotation);
+		}
+
+		for (ticker_name, quote) in quotes {
+			if let Some(asset) = ticker_to_asset_map.get(ticker_name.as_str()) {
+				let symbol = asset.symbol.clone();
+				// We use the bid price as the price
+				let price = quote.b;
```
Comment (gianfra-t) [inline dia-batching-server/src/api/polygon.rs:72]: Do you know what is the price that the previous endpoint `conversions` was using? Because the bid price can sometimes be quite lower from the actual previous trade price.

**Review by gianfra-t — APPROVED** [review]
Looks good to me too 👍. About the comment I left, it's more about trading. Perhaps using the `bid` price (or `ask`) in this scenario is sufficient since these pairs must be very liquid and the gap is probably low. 

Sadly `/v1/conversion/` doesn't specify what they use so at least here we know what is the price, exactly.

File: dia-batching-server/src/api/polygon.rs (hunk 16 above)
Comment (ebma) [reply dia-batching-server/src/api/polygon.rs:72]: I compared the responses from the 'conversions' and the 'all-tickers' and it seemed to me like 'conversions' returned the bid price but rounded. 

See 
https://api.polygon.io/v1/conversion/AUD/USD?amount=1&precision=4&apiKey={API_KEY}
vs
https://api.polygon.io/v2/snapshot/locale/global/markets/forex/tickers?tickers=C:AUDUSD&apiKey={API_KEY}
